File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int length = nums.size();

        vector<int> post_prod(length,1);
        int post = 1;
        for (int i = length-1; i>=0; i--){
            post_prod[i] = post;
            post *= nums[i];
        }

        vector<int> res;
        int prev = 1;
        for (int i = 0; i < length; ++i) {
            res.push_back(post_prod[i] * prev);
            prev *= nums[i];
        }

        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        post = 1
        post_prod = [1]*len(nums)
        for i in range(len(nums)-1,-1,-1):
            post_prod[i] = post
            post *= nums[i]

        res = []
        prev = 1
        for i in range(len(nums)):
            res.append(post_prod[i]*prev)
            prev *= nums[i]
        
        return res
       
        
